typed/deduced.go

```diff
@@ -17,162 +17,28 @@ limitations under the License.
-}
-
-func mergeDeduced(lhs, rhs value.Value) value.Value {
-	// If both sides are maps, merge them, otherwise return right
-	// side.
-	if rhs.MapValue == nil || lhs.MapValue == nil {
-		return rhs
-	}
-
-	v := value.Value{MapValue: &value.Map{}}
-	for i := range lhs.MapValue.Items {
-		child := lhs.MapValue.Items[i]
-		v.MapValue.Set(child.Name, child.Value)
-	}
-	for i := range rhs.MapValue.Items {
-		child := rhs.MapValue.Items[i]
-		if sub, ok := v.MapValue.Get(child.Name); ok {
-			new := mergeDeduced(sub.Value, child.Value)
-			v.MapValue.Set(child.Name, new)
-		} else {
-			v.MapValue.Set(child.Name, child.Value)
-		}
-	}
-	return v
-}
-
-func (dv deducedTypedValue) Compare(rhs TypedValue) (c *Comparison, err error) {
-	trhs, ok := rhs.(deducedTypedValue)
-	if !ok {
-		return nil, errorFormatter{}.
-			errorf("can't merge deducedTypedValue with %T", rhs)
-	}
-
-	c = &Comparison{
-		Removed:  fieldpath.NewSet(),
-		Modified: fieldpath.NewSet(),
-		Added:    fieldpath.NewSet(),
-	}
-
-	added(dv.value, trhs.value, fieldpath.Path{}, c.Added)
-	added(trhs.value, dv.value, fieldpath.Path{}, c.Removed)
-	modified(dv.value, trhs.value, fieldpath.Path{}, c.Modified)
-
-	merge, err := dv.Merge(rhs)
-	if err != nil {
-		return nil, err
-	}
-	c.Merged = merge
-	return c, nil
-}
-
-func added(lhs, rhs value.Value, path fieldpath.Path, set *fieldpath.Set) {
-	if lhs.MapValue == nil && rhs.MapValue == nil {
-		// Both non-maps, nothing added, do nothing.
-	} else if lhs.MapValue == nil && rhs.MapValue != nil {
-		// From leaf to map, add leaf fields of map.
-		fieldsetDeduced(rhs, path, set)
-	} else if lhs.MapValue != nil && rhs.MapValue == nil {
-		// Went from map to field, add field.
-		set.Insert(path)
-	} else {
-		// Both are maps.
-		for i := range rhs.MapValue.Items {
-			child := rhs.MapValue.Items[i]
-			np := path.Copy()
-			np = append(np, fieldpath.PathElement{FieldName: &child.Name})
-
-			if v, ok := lhs.MapValue.Get(child.Name); ok {
-				added(v.Value, child.Value, np, set)
-			} else {
-				fieldsetDeduced(child.Value, np, set)
-			}
-		}
-	}
-}
-
-func modified(lhs, rhs value.Value, path fieldpath.Path, set *fieldpath.Set) {
-	if lhs.MapValue == nil && rhs.MapValue == nil {
-		if !reflect.DeepEqual(lhs, rhs) {
-			set.Insert(path)
-		}
-	} else if lhs.MapValue != nil && rhs.MapValue != nil {
-		// Both are maps.
-		for i := range rhs.MapValue.Items {
-			child := rhs.MapValue.Items[i]
-
-			v, ok := lhs.MapValue.Get(child.Name)
-			if !ok {
-				continue
-			}
-
-			np := path.Copy()
-			np = append(np, fieldpath.PathElement{FieldName: &child.Name})
-			modified(v.Value, child.Value, np, set)
-		}
-	}
-}
-
-// RemoveItems does nothing because all lists in a deducedTypedValue are considered atomic,
-// and there are no maps because it is indistinguishable from a struct.
-func (dv deducedTypedValue) RemoveItems(_ *fieldpath.Set) TypedValue {
-	return dv
+	return AsTypedUnvalidated(v, &deducedSchema, deducedName)
```

Comment:
Since we're getting rid of all of this, I'm curious if we should make TypedValue back into a struct rather than an interface.

Reply:
Will that still be compatible with what we did in k8s? I think yes but I'm not sure

Reply:
I know that Daniel wants to start keeping the thing compatible. It's not done yet, let's use that chance ;-)